src/server/json_family.cc

```diff
@@ -773,46 +782,63 @@ OpResult<long> OpClear(const OpArgs& op_args, string_view key, string_view path)
   return clear_items;
 }
 
-// Returns string vector that represents the pop out values.
-OpResult<vector<OptString>> OpArrPop(const OpArgs& op_args, string_view key, string_view path,
-                                     int index) {
-  vector<OptString> vec;
-  auto cb = [&](const auto& path, JsonType& val) {
-    if (!val.is_array() || val.empty()) {
-      vec.emplace_back(nullopt);
-      return;
-    }
+void ArrayPop(std::optional<std::string_view>, int index, JsonType* val,
+              vector<OptString>* result) {
+  if (!val->is_array() || val->empty()) {
+    result->emplace_back(nullopt);
+    return;
+  }
 
-    size_t removal_index;
-    if (index < 0) {
-      int temp_index = index + val.size();
-      removal_index = abs(temp_index);
-    } else {
-      removal_index = index;
-    }
+  size_t removal_index;
+  if (index < 0) {
+    int temp_index = index + val->size();
+    removal_index = abs(temp_index);
+  } else {
+    removal_index = index;
+  }
 
-    if (removal_index >= val.size()) {
-      removal_index %= val.size();  // rounded to the array boundaries.
-    }
+  if (removal_index >= val->size()) {
+    removal_index %= val->size();  // rounded to the array boundaries.
+  }
 
-    auto it = std::next(val.array_range().begin(), removal_index);
-    string str;
-    error_code ec;
-    it->dump(str, {}, ec);
-    if (ec) {
-      VLOG(1) << "Failed to dump JSON to string with the error: " << ec.message();
-      return;
-    }
+  auto it = val->array_range().begin() + removal_index;
+  string str;
+  error_code ec;
+  it->dump(str, {}, ec);
+  if (ec) {
+    LOG(ERROR) << "Failed to dump JSON to string with the error: " << ec.message();
+    result->emplace_back(nullopt);
+    return;
+  }
 
-    vec.push_back(str);
-    val.erase(it);
-  };
+  result->push_back(std::move(str));
+  val->erase(it);
+  return;
+};
 
-  OpStatus status = UpdateEntry(op_args, key, path, cb);
+// Returns string vector that represents the pop out values.
+OpResult<vector<OptString>> OpArrPop(const OpArgs& op_args, string_view key, string_view path,
+                                     int index, JsonPathV2 expression) {
+  vector<OptString> vec;
+  OpStatus status;
+  if (std::holds_alternative<json::Path>(expression)) {
+    const json::Path& json_path = std::get<json::Path>(expression);
+    auto cb = [&vec, index](optional<string_view>, JsonType* val) {
+      ArrayPop(nullopt, index, val, &vec);
+      return false;
+    };
+    status = UpdateEntry(op_args, key, json_path, std::move(cb));
+  } else {
+    auto cb = [&](const auto& path, JsonType& val) {
+      ArrayPop(nullopt, index, &val, &vec);
+      return false;
+    };
+
+    status = UpdateEntry(op_args, key, path, cb);
+  }
```

Comment:
please consider to make one callback, because they are the same and you can use some type deduction to create one common type for JsonType

Reply:
these c++ gymnastics here are  not worth it - the code is transitional and we will remove JsonExpression once everyone will be handled by json::Path spend some time released